### PR TITLE
minor code cleanup from multicore merge comments

### PR DESCRIPTION
**Pull request by Octachron**
This PR implements two minor cleanup proposed in #10861 that seemed too small to require either new issues or separate PRs:

- `caml_realloc_global` is simplified since there are no read barriers
- `Obj.is_shared` and `Obj.compare_and_swap_field` are removed from the `Obj` module

(Note that I am not proposing to backport the changes to 5.0)

### CODE REVIEW

**Comment by kayceesrk** [comment]
LGTM.

**Comment by Octachron** [comment]
@kayceesrk should I wait for an official approval?

**Review by kayceesrk — APPROVED** [review]

**Comment by kayceesrk** [comment]
There you go :-)

**Pull request merged**
